pkg/ztunnel/xds/stream_processor.go

```diff
@@ -0,0 +1,349 @@
+	"github.com/cilium/cilium/pkg/option"
+)
+
+type StreamProcessorParams struct {
+	// The backing gRPC bidi stream initiated by zTunnel
+	Stream v3.AggregatedDiscoveryService_DeltaAggregatedResourcesServer
+	// Channel where the StreamProcessor listens for new DeltaDiscoveryRequest.
+	// this is fed directly from Stream.
+	StreamRecv chan *v3.DeltaDiscoveryRequest
+	// Channel where the StreamProcessor listens for Endpoint event.
+	// this is fed by subscribing to EndpointManager.
+	EndpointEventRecv chan *EndpointEvent
+	// Reference to agent's CEP watcher and resource cache.
+	// This will be the source of truth for serving WDS API.
+	K8sCiliumEndpointsWatcher *watchers.K8sCiliumEndpointsWatcher
+	Log                       *slog.Logger
+}
+
+// StreamProcessor implements the logic for handling xDS streams to zTunnel.
+// It is abstracted away from the primary stream initialization handler to
+// promote decoupling and the handling multiple streams without a shared set of
+// channels being required on the Server object.
+type StreamProcessor struct {
+	stream         v3.AggregatedDiscoveryService_DeltaAggregatedResourcesServer
+	streamRecv     chan *v3.DeltaDiscoveryRequest
+	endpointRecv   chan *EndpointEvent
+	expectedNonce  map[string]struct{}
+	log            *slog.Logger
+	endpointSource EndpointEventSource
+}
+
+// EndpointSource provides data for XDS server from different data sources in the agent.
+type EndpointSource struct {
+	k8sCiliumEndpointsWatcher *watchers.K8sCiliumEndpointsWatcher
+	sp                        *StreamProcessor
+}
+
+func NewEndpointSource(k *watchers.K8sCiliumEndpointsWatcher, sp *StreamProcessor) *EndpointSource {
+	return &EndpointSource{
+		k8sCiliumEndpointsWatcher: k,
+		sp:                        sp,
+	}
+}
+
+// Interface different data sources need to implement for usage with StreamProcessor.
+type EndpointEventSource interface {
+	// Subscribe to Endpoint events.
+	SubscribeToEndpointEvents(ctx context.Context, wg *sync.WaitGroup)
+	ListAllEndpoints(ctx context.Context) ([]*types.CiliumEndpoint, error)
+}
+
+func NewStreamProcessor(params *StreamProcessorParams) *StreamProcessor {
+	sp := &StreamProcessor{
+		stream:        params.Stream,
+		streamRecv:    params.StreamRecv,
+		endpointRecv:  params.EndpointEventRecv,
+		log:           params.Log,
+		expectedNonce: make(map[string]struct{}),
+	}
+	sp.endpointSource = NewEndpointSource(params.K8sCiliumEndpointsWatcher, sp)
+	return sp
+}
+
+func (es *EndpointSource) SubscribeToEndpointEvents(ctx context.Context, wg *sync.WaitGroup) {
+	if option.Config.EnableCiliumEndpointSlice {
+		newSliceEvents := es.k8sCiliumEndpointsWatcher.GetCiliumEndpointSliceResource().Events(ctx, resource.WithErrorHandler(resource.AlwaysRetry))
+		for e := range newSliceEvents {
+			if e.Object == nil {
+				e.Done(nil)
+				continue
+			}
+			if e.Kind == resource.Sync {
+				wg.Done()
+				e.Done(nil)
+				continue
+			}
+			for _, coreCep := range e.Object.Endpoints {
```

Comment:
That won't be as easy as this unfortunately. 
You can get `upsert` event for CES that deletes CEP in CES.
For example CES `[ep1, ep2]` gets updated to CES `[ep1]`

Reply:
Looks reasonable, thanks!